server/src/main/java/org/elasticsearch/action/admin/indices/readonly/AddIndexBlockResponse.java

```diff
@@ -0,0 +1,274 @@
+        for (AddBlockResult index : indices) {
+            index.toXContent(builder, params);
+        }
+        builder.endObject();
+    }
+
+    @Override
+    public String toString() {
+        return Strings.toString(this);
+    }
+
+    public static class AddBlockResult implements Writeable, ToXContentFragment {
+
+        private final Index index;
+        private final @Nullable Exception exception;
+        private final @Nullable AddBlockShardResult[] shards;
+
+        public AddBlockResult(final Index index) {
+            this(index, null, null);
+        }
+
+        public AddBlockResult(final Index index, final Exception failure) {
+            this(index, Objects.requireNonNull(failure), null);
+        }
+
+        public AddBlockResult(final Index index, final AddBlockShardResult[] shards) {
+            this(index, null, Objects.requireNonNull(shards));
+        }
+
+        private AddBlockResult(final Index index, @Nullable final Exception exception, @Nullable final AddBlockShardResult[] shards) {
+            this.index = Objects.requireNonNull(index);
+            this.exception = exception;
+            this.shards = shards;
+        }
+
+        AddBlockResult(final StreamInput in) throws IOException {
+            this.index = new Index(in);
+            this.exception = in.readException();
+            this.shards = in.readOptionalArray(AddBlockShardResult::new, AddBlockShardResult[]::new);
+        }
+
+        @Override
+        public void writeTo(final StreamOutput out) throws IOException {
+            index.writeTo(out);
+            out.writeException(exception);
+            out.writeOptionalArray(shards);
+        }
+
+        public Index getIndex() {
+            return index;
+        }
+
+        public Exception getException() {
+            return exception;
+        }
+
+        public AddBlockShardResult[] getShards() {
+            return shards;
+        }
+
+        public boolean hasFailures() {
+            if (exception != null) {
+                return true;
+            }
+            if (shards != null) {
+                for (AddBlockShardResult shard : shards) {
+                    if (shard.hasFailures()) {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        @Override
+        public XContentBuilder toXContent(final XContentBuilder builder, final Params params) throws IOException {
+            builder.startObject(index.getName());
+            {
+                if (hasFailures()) {
+                    builder.field("blocked", false);
+                    if (exception != null) {
+                        builder.startObject("exception");
+                        ElasticsearchException.generateFailureXContent(builder, params, exception, true);
+                        builder.endObject();
+                    } else {
+                        builder.startObject("failedShards");
```

Comment:
I think this should be `failed_shards` rather than camel case to match our usual API conventions